### ISSUE TITLE
Extract styles?

### ISSUE THREAD

**Issue by webchaz** [styled-components/styled-components#1018]
I'm having a hard time finding any information on how to extract the inline styles to a separate file.  Is this possible? Using extract-text-plugin in webpack or similar.

Thanks

**Comment by philpl** [comment]
This is a tougher question to answer, but let me break it down :smile:

We don't support static CSS extraction. You can try out emotion which does.

You might not need static CSS extraction actually, because of several reasons:

- There's SSR which sends only critical CSS, instead of all static CSS, for the entire page. You don't even need to do SSR, but can use snapshotting (react-snapshot) or generate a static page (gatsby, et al), which basically saves a SSR result to a html.
- Static extraction doesn't generate dynamic CSS, which means your page will either appear broken until the JS executes, or you'll need to defer until the JS is loaded
- Caching doesn't actually buy you an advantage, because the JS bundles will likely always change in tandem with the extracted CSS

In v3 we will add preprocessing, which will actually a bigger advantage. As part of that we might support an option for static extraction, since the core library that will bring preprocessing will probably also be integrated into emotion, which does support extraction. So it might become an option. Or not :wink:

I hope that answers your question.

Oh, **bonus round** while I'm at it. We won't support it in SC's standard form, because it doesn't work with our way of writing CSS. As mixins and static and dynamic CSS all have no specific specificity (all have the same importance), we order your CSS in the order you write it. This doesn't map at all to static extraction: https://gist.github.com/philpl/7b8a021f19ead2ec534091976a1ff6dc#file-2-specificity-md (This document explains way based on a simple example)

Emotion achieves static extraction by breaking with our previous assumptions and way of generating CSS.

**Comment by webchaz** [comment]
Thanks @philpl for the explanation and suggestions.    It seems to be appending class names to `styled-components` instead of replacing old ones.  Any reason for this?  Basically for a simple button component that changes based on props, this seems to just grow infinitely.

<img width="1079" alt="screenshot 2017-07-20 19 13 45" src="https://user-images.githubusercontent.com/1342709/28442676-9227b97a-6d7f-11e7-8f8c-54d0b87f6bc6.png">

**Comment by philpl** [comment]
For every different CSS string per styled component there'll be a new class hash. As it's really unrealistic to reach a high amount of styles in non-artificial conditions, no one has complained about reaching a performance limit yet. Albeit there being one at the tens of thousands, where css injection slows down to a stunningly low speed.

But we have a fix for our comparably slow injection speed already. It's solved by using batching. Other libs use the speedy api which prevent you from editing css in production due to it. So we'll try batching first, which seems to increase the theoretical limit by a lot more again.

So to sum this up: Totally unrelated to the original question (No worries :wink:) and not a big deal

**Comment by webchaz** [comment]
@philpl Ha, Thank you - Yes I hijacked the same question.    That sounds great though, loving styled components so far.  Thanks again for detailed responses.

**Comment by Undistraction** [comment]
@kitten Above you say:

> In v3 we will add preprocessing

Is this still being considered? Is preprocessing on the roadmap?

**Comment by sunOpar** [comment]
When uses `create-react-app`, the ExtractTextPlugin will enable default. How to make them well together.

**Comment by xinzhang** [comment]
@webchaz this inline styles actually an issue for us especially when we running the web site in cloud and they charge by egress network data size. we have a unzipped 1.5MB html page and 1.3MB styles inline, it is gzipped as 200KB but it is still too large. Is there any suggestion what we can do? By the way, we are using Next and calling sheet.getStyleElement() to have all the css which is what Next needed.

**Comment by kitten** [comment]
@xinzhang That does sound like quite the egress issue. I assume that means you're server rendering all pages? As it so happens, I think there's a way to export the SSR'd styles into an external CSS file, which of course only makes sense if you're prerendering your pages and are statically exporting them. Are you CDN caching your pages?

**Comment by xinzhang** [comment]
@kitten ,  I think there's a way to export the SSR'd styles into an external CSS file --> as we do ssr, this is really what we need. Currently the code in _Document is like below,
```
static async getInitialProps(ctx) {
    const sheet = new ServerStyleSheet();
    const page = ctx.renderPage(App => props => sheet.collectStyles(<App {...props} />));
    const styleTags = sheet.getStyleElement();
}
```

Then in the render() function, we have to include the styleTags in the html.

```
render() {
    return (
      <html lang="en">
        <Head>
          {this.props.styleTags}
        </Head>
        <body>
          <Main />
        </body>
     </html>
  )
}
```
I wonder if we can output this styleTags into a separate file and I don't know how. because this is the whole part of the process of the Next framework to return a SSR page.

**Comment by baba43** [comment]
I just discovered this issue because I'm looking for a way to use styled components in a WordPress plugin based on React. For this to work properly, it would be necessary to generate all the necessary styles as a separate file.

Just to list another use case.

**Comment by FerhatAvdic** [comment]
React testing library needs the stylesheet attached to the DOM within a test to be able to check for applied styles via "toHaveStyle" method.
https://github.com/testing-library/jest-dom/issues/295#issuecomment-957865299

**Comment by drianoaz** [comment]
Are there any plans to reconsider adding of Static CSS Extraction? 

This might be a good way to support React 18 Streaming SSR #3658

**Comment by hgcsm** [comment]
Any update here?  We have same issue where our pages are having trouble getting indexed due to excessively large html file size, and a low text to code ratio due to the css..
